subprojects/core/src/main/java/org/gradle/api/internal/CompositeDomainObjectSet.java

```diff
@@ -163,4 +145,135 @@ public void all(Action<? super T> action) {
+            }
+            return false;
+        }
+
+        @Override
+        public int size() {
+            int size = 0;
+            for (DomainObjectCollection<? extends T> ts : store) {
+                size += ts.size();
+            }
+            return size;
+        }
+
+        @Override
+        public boolean isEmpty() {
+            for (DomainObjectCollection<? extends T> ts : store) {
+                if (!ts.isEmpty()) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        @Override
+        public boolean contains(Object o) {
+            for (DomainObjectCollection<? extends T> ts : store) {
+                if (ts.contains(o)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public Iterator<T> iterator() {
+            if (store.isEmpty()) {
+                return Iterators.emptyIterator();
+            }
+            if (store.size() == 1) {
+                return (Iterator<T>) store.get(0).iterator();
+            }
+            List<Iterator<? extends T>> iterators = new ArrayList<Iterator<? extends T>>(store.size());
+            for (DomainObjectCollection<? extends T> ts : store) {
+                iterators.add(ts.iterator());
+            }
+            return Iterators.<T>concat(iterators.toArray(new Iterator[0]));
+        }
+
+        @Override
+        public Object[] toArray() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public <V> V[] toArray(V[] a) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean add(T t) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean remove(Object o) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean containsAll(Collection<?> c) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends T> c) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean removeAll(Collection<?> c) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean retainAll(Collection<?> c) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void clear() {
+            throw new UnsupportedOperationException();
+        }
+
+        public void addComposited(DomainObjectCollection<? extends T> collection) {
+            this.store.add(collection);
+        }
+
+        public void removeComposited(DomainObjectCollection<? extends T> collection) {
+            Iterator<DomainObjectCollection<? extends T>> iterator = store.iterator();
```

Comment:
Because we use identity comparison, not `equals`. We're removing a list from a list, and we don't want to pay the price of comparing the lists elements, knowing that this is a composition of lists where only identity matters.